pkg/operator/imageregistrycertificates.go

```diff
@@ -0,0 +1,115 @@
+	openshiftConfigLister corev1listers.ConfigMapNamespaceLister
+
+	cachesToSync []cache.InformerSynced
+	queue        workqueue.RateLimitingInterface
+}
+
+func NewImageRegistryCertificatesController(
+	coreClient corev1client.CoreV1Interface,
+	configMapInformer corev1informers.ConfigMapInformer,
+	serviceInformer corev1informers.ServiceInformer,
+	imageConfigInformer configv1informers.ImageInformer,
+	openshiftConfigInformer corev1informers.ConfigMapInformer,
+) *ImageRegistryCertificatesController {
+	c := &ImageRegistryCertificatesController{
+		coreClient:            coreClient,
+		configMapLister:       configMapInformer.Lister().ConfigMaps(defaults.ImageRegistryOperatorNamespace),
+		serviceLister:         serviceInformer.Lister().Services(defaults.ImageRegistryOperatorNamespace),
+		imageConfigLister:     imageConfigInformer.Lister(),
+		openshiftConfigLister: openshiftConfigInformer.Lister().ConfigMaps(defaults.OpenShiftConfigNamespace),
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ImageRegistryCertificatesController"),
+	}
+
+	configMapInformer.Informer().AddEventHandler(c.eventHandler())
+	c.cachesToSync = append(c.cachesToSync, configMapInformer.Informer().HasSynced)
+
+	serviceInformer.Informer().AddEventHandler(c.eventHandler())
+	c.cachesToSync = append(c.cachesToSync, serviceInformer.Informer().HasSynced)
+
+	imageConfigInformer.Informer().AddEventHandler(c.eventHandler())
+	c.cachesToSync = append(c.cachesToSync, imageConfigInformer.Informer().HasSynced)
+
+	openshiftConfigInformer.Informer().AddEventHandler(c.eventHandler())
+	c.cachesToSync = append(c.cachesToSync, openshiftConfigInformer.Informer().HasSynced)
+
+	return c
+}
+
+func (c *ImageRegistryCertificatesController) eventHandler() cache.ResourceEventHandler {
+	const workQueueKey = "instance"
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.queue.Add(workQueueKey) },
+		UpdateFunc: func(old, new interface{}) { c.queue.Add(workQueueKey) },
+		DeleteFunc: func(obj interface{}) { c.queue.Add(workQueueKey) },
+	}
+}
+
+func (c *ImageRegistryCertificatesController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *ImageRegistryCertificatesController) processNextWorkItem() bool {
+	obj, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	klog.V(1).Infof("get event from workqueue")
+	if err := c.sync(); err != nil {
+		c.queue.AddRateLimited(workqueueKey)
+		klog.Errorf("ImageRegistryCertificatesController: unable to sync: %s, requeuing", err)
+	} else {
+		c.queue.Forget(obj)
+		klog.Infof("ImageRegistryCertificatesController: event from workqueue successfully processed")
+	}
+	return true
+}
+
+func (c *ImageRegistryCertificatesController) sync() error {
+	g := resource.NewGeneratorCAConfig(c.configMapLister, c.imageConfigLister, c.openshiftConfigLister, c.serviceLister, c.coreClient)
```

Comment:
Could this be done when creating the object on `NewImageRegistryCertificatesController()` ?

Reply:
Generators are lightweight objects (i.e. their constructors are trivial) and in this case it should be created on stack, so it doesn't matter. Eventually I'd expect us to get rid of these generators and inline them into controllers.